atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java

```diff
@@ -392,6 +430,40 @@ public RangeMap<LightweightOppToken, ImmutableSet<CassandraServer>> getTokenMap(
         return tokenMap;
     }
 
+    private void setTokenMap(ImmutableRangeMap<LightweightOppToken, ImmutableSet<CassandraServer>> tokenRing) {
+        // intern to deduplicate equivalent rings to avoid memory overhead of different views
+        ImmutableRangeMap<LightweightOppToken, ImmutableSet<CassandraServer>> newRing =
+                tokenRingInterner.intern(tokenRing);
+        ImmutableRangeMap<LightweightOppToken, ImmutableSet<CassandraServer>> previousRing = tokenMap; // volatile read
+
+        @SuppressWarnings("ReferenceEquality") // interning provides ability to rely on reference equality for changes
+        boolean ringChanged = previousRing != newRing;
+        if (ringChanged) {
```

Comment:
Do we have any tests for this?

Reply:
in fact I think we should break up this PR into 2
1. race condition + tests
2. 2. refactor + tests

Reply:
will split this up when I have cycles